### PR TITLE
feat: Add support DataTemplate is container host

### PR DESCRIPTION
**Pull request by jeromelaban**
GitHub Issue (If applicable): fixes https://github.com/unoplatform/uno/issues/2477

## PR Type

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs [hunk 1]
```diff
@@ -1115,13 +1101,53 @@ internal DependencyObject GetContainerForIndex(int index)
...
+			}
+			else if (isContainerFromTemplateRoot)
+			{
+				container = elementTemplateRoot;
+				SetContainerContent(container, item, isGeneratedContainer: true);
+			}
+			else
+			{
+				container = GetContainerForItemOverride();
+				SetContainerContent(container, elementTemplateRoot);
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs:1128]: You still seem to be setting `elementTemplateRoot` as the content here. I would've thought you'd want to ignore `elementTemplateRoot` on this code path.

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs:1128]: That's the issue I was looking for :) There are a bunch of screenshot differences for which there are no associatd unit tests, I'll add a few for that case.

Thanks!

File: src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs [hunk 2]
```diff
@@ -1109,21 +1108,46 @@ internal void PrepareContainerForIndex(DependencyObject container, int index)
...
 		{
 			var item = ItemFromIndex(index);
 
-			var container = IsItemItsOwnContainerOverride(item)
-				? item as DependencyObject
-				: GetContainerForItemOverride();
+			if (IsItemItsOwnContainerOverride(item))
+			{
+				var container = item as DependencyObject;
+				container.SetValue(ItemsControlForItemContainerProperty, new WeakReference<ItemsControl>(this));
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs:1121]: You should be able to use `IWeakReferenceProvider.WeakReference` here, but maybe that's for another PR :)

File: src/Uno.UI/UI/Xaml/Controls/Primitives/Selector.cs [hunk 3]
```diff
@@ -28,6 +28,11 @@ public partial class Selector : ItemsControl
 		/// </summary>
 		internal virtual bool IsSingleSelection => true;
 
+		/// <summary>
+		/// Templates for which it's known that the template root doesn't qualify as a container.
+		/// </summary>
+		private readonly HashSet<DataTemplate> _itemTemplatesThatArentContainers = new HashSet<DataTemplate>();
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Primitives/Selector.cs:34]: Should we clear this if the `ItemTemplateSelector` changes ?

File: src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs [hunk 4]
```diff
@@ -1,4 +1,6 @@
-﻿#if NETSTANDARD
+﻿#nullable enable
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs:1]: 🎉🎉🎉🎉

File: src/Uno.UI/UI/Xaml/Controls/Primitives/Selector.cs (hunk 3 above)
Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/Controls/Primitives/Selector.cs:34]: Yes! Fixed

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ListViewBase.cs [hunk 5]
```diff
@@ -4,13 +4,38 @@
 using System.Text;
 using System.Threading.Tasks;
 using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Private.Infrastructure;
+#if NETFX_CORE
+using Uno.UI.Extensions;
+#elif __IOS__
+using UIKit;
```
Comment (davidjohnoliver) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ListViewBase.cs:11]:
```suggestion
using UIKit;
#elif __MACOS__
using AppKit;
```

**Pull request merged**